### ISSUE TITLE
Support to  Analytic-DPM and DPM-Solver

### ISSUE THREAD

**Issue by lhaippp** [lucidrains/denoising-diffusion-pytorch#141]
Hi Phil, thanks again for the great job!

Recently, Analytic-DPM and DPM-Solver, proposed optimal analytical solutions and better sampling methods, respectively.
Would you mind considering adding them to this repo?

Best regards

**Comment by lucidrains** [comment]
@lhaippp hey thanks Li! hmm, are you sure dpm-solver works? if it does work, wouldn't stable diffusion be using it?

**Comment by lucidrains** [comment]
@lhaippp famous last words

the majority of papers don't live up to their lofty claims

**Comment by lucidrains** [comment]
@lhaippp ok, you find a couple researchers or deep learning practitioners that can give good testimonies and i'll look into it

**Comment by lucidrains** [comment]
@lhaippp i'm in touch with the people behind stable diffusion, and none of them told me about dpm solver, and usually they are on top of things

**Comment by lucidrains** [comment]
@lhaippp verifying is a good idea :) most papers do not end up working as advertised, or has some catch

**Comment by lucidrains** [comment]
@lhaippp hey Li, i actually heard from someone i trust that dpm solver

will look into it this week

**Comment by lhaippp** [comment]
@lucidrains hi Phil, that's awesome! Look forward

**Comment by Mut1nyJD** [comment]
I've got an implementation of DPM-Solver++ in combo with Elucidated Diffusion working with this repo. If interested I can make a pull-request

**Comment by lucidrains** [comment]
@Mut1nyJD omg, that would be amazing! so dpm-solver++ is working for you then?

**Comment by lucidrains** [comment]
@Mut1nyJD i'm confused, how could dpm-solver++ be compatible with elucidated sampling?

**Comment by Mut1nyJD** [comment]
> @Mut1nyJD omg, that would be amazing! so dpm-solver++ is working for you then?

Yeah it works at least on the model I've trained.  Katherine Crowson made it compatible with Elucidated sampling what she calls k-diffusion (k for Karras). 

I will clean it up and submit a pull request.

**Comment by lucidrains** [comment]
@Mut1nyJD thanks for relaying this! ugh, i'm missing out not being on discord

**Comment by Mut1nyJD** [comment]
@lucidrains  

No problem you're welcome! It is hard to keep up with all the developments. 

What I am personal curious is would be the addition of flash attention do you think that would work? And would it have any benefit to linear attention that is mostly used right now

And the other one is Self-Attention-Guidance (https://github.com/PnDong/Self-Attention-Guidance)

**Comment by lucidrains** [comment]
@Mut1nyJD yup, flash attention will be better than linear attention, although you will still face quadratic compute price

without reading the paper, is self attention guidance for better binding to text?

**Comment by Mut1nyJD** [comment]
@lucidrains Yeah but should be still faster than full attention? Is your flash attention mature enough to try?

Well I guess you could also use the cross attention I am sure they are working on that but this paper/implementation is primarily about the general self attention layers so as far as I understood it is for unsupervised and maybe class supervised diffusion

**Comment by lucidrains** [comment]
@Mut1nyJD yup, under certain conditions it is faster

my flash attention repository is using cosine similarity, and unfortunately Robin had some failed experiments trying it out, i can't recommend it until more validation is done. good news is that flash attention will eventually be offered in pytorch! it will probably mature by v2.1

oh! 🤦 yes i got confused

just scanned the diagrams of the paper and i like it! thank you for mentioning it, may try it out when i find time

**Comment by lucidrains** [comment]
@Mut1nyJD what would happen if self attention guidance were applied to a language model?

**Comment by Mut1nyJD** [comment]
@lucidrains 

> my flash attention repository is using cosine similarity, and unfortunately Robin had some failed experiments using it, so i can't recommend it until more validation is done. good news is that flash attention will eventually be offered in pytorch!

Ok thank you for the feedback. I will have a look at the official pytorch implementation. Although I due tend to be back a version from current main release 😄 

> oh! 🤦 yes i got confused
> 
> just scanned the diagrams of the paper and i like it! thank you for mentioning it, may try it out when i find time

Yeah looks like an interesting not too expansive improvement. That's great I will try my best as well to decipher it let you know if I got something if you haven't beaten me by then 😄

**Comment by lucidrains** [comment]
closing because of https://github.com/lucidrains/denoising-diffusion-pytorch/pull/148
